text/0000-restrict-constants-in-patterns.md

```diff
@@ -0,0 +1,568 @@
+are. The advantages and disadvantages are discussed above.
+
+**Embrace semantic equality.** We could opt to just go straight
+towards "semantic equality". Howver, it seems better to reset the
+semantics to a base point that everyone can agree on, and then extend
+from that base point. Moreover, adopting semantic equality straight
+out would be a riskier breaking change, as it could silently change
+the semantics of existing programs (whereas the current proposal only
+causes compilation to fail, never changes what an existing program
+will do).
+
+# Discussion thread summary
+
+This section summarizes various points that were raised in the
+[internals thread] which are related to patterns but didn't seem to
+fit elsewhere.
+
+**Overloaded patterns.** Some languages, notably Scala, permit
+overloading of patterns. This is related to "semantic equality" in
+that it involves executing custom, user-provided code at compilation
+time.
+
+**Pattern synonyms.** Haskell offers a feature called "pattern
+synonyms" and
+[it was argued](https://internals.rust-lang.org/t/how-to-handle-pattern-matching-on-constants/2846/39?u=nikomatsakis)
+that the current treatment of patterns can be viewed as a similar
+feature. This may be true, but constants-in-patterns are lacking a
+number of important features from pattern synonyms, such as bindings,
+as
+[discussed in this response](https://internals.rust-lang.org/t/how-to-handle-pattern-matching-on-constants/2846/48?u=nikomatsakis).
+The author feels that pattern synonyms might be a useful feature, but
+it would be better to design them as a first-class feature, not adapt
+constants for that purpose.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+**Should we also adjust the exhaustiveness and match analysis
+algorithm to be more conservative around constants?** This RFC just
+proposes limiting the types of constants that can be used in a match
+pattern. However, since the code currently inlines the actual values
+of constants before doing exhaustiveness checking, this also implies
+that it can compute exhaustiveness and dead-code in cases where it
+arguably should not be able to.
+
+For example, the following code
+[fails to compile](http://is.gd/PJjNKl) because it contains dead-code:
+
+```rust
+const X: u64 = 0;
+const Y: u64 = 0;
+fn bar(foo: u64) {
+    match foo {
+        X => { }
+        Y => { }
+        _ => { }
+    }
+}
+```
+
+However, we would be unable to perform such an analysis in a more
+generic context, such as with an associated constant:
+
+```rust
+trait Trait {
+    const X: u64;
+    const Y: u64;
+}
+
+fn bar<T:Trait>(foo: u64) {
+    match foo {
+        T::X => { }
+        T::Y => { }
+        _ => { }
+    }
+}
+```
+
+Here, although it may well be that `T::X == T::Y`, we can't know for
+sure. So, for consistency, we may wish to treat all constants opaquely
+regardless of whether we are in a generic context or not.
```

Comment:
Alternatively, in the future, we could forbid this kind of matching if it cannot be guaranteed that `T::X != T::Y` (which could be generically guaranteed by allowing such a statement in where clauses)